### PR TITLE
dev-cmd/audit: allow checksums to change when switching to/from a PyPI url

### PR DESCRIPTION
**Pull request by dtrodrigues**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew tests` with your changes locally?
- [x] Have you successfully run `brew man` locally and committed any changes?

-----
When possible for Python packages, it's preferable to use the PyPI url vs GitHub/other urls to allow for automatic resource bumps. However, in general, the checksum of the GitHub archive will not match the checksum of the PyPI source package. This means that the audit will complain that the sha256 changed without changing the version if a PR is created which only switches the URL to/from PyPI. 

Supports https://github.com/Homebrew/homebrew-core/pull/64243

### CODE REVIEW

**Review by AdrianMichigan — APPROVED** [review]

File: Library/Homebrew/dev-cmd/audit.rb
Comment (Rylan12) [inline Library/Homebrew/dev-cmd/audit.rb:925]: Do we need to restrict this to PyPI URLs? I would think that switching URLs any time might mean a different sha256

If so, how about:

```suggestion
         current_checksum != newest_committed_checksum &&
         newest_committed_url.start_with?(pypi_prefix) != current_url.start_with?(pypi_prefix)
```

That might be just a personal preference but I think that's easier to read than negating an xor.

File: Library/Homebrew/dev-cmd/audit.rb
Comment (Rylan12) [inline Library/Homebrew/dev-cmd/audit.rb:921]:
```suggestion
      pypi_prefix = PyPI:: PYTHONHOSTED_URL_PREFIX
```

**Review by Rylan12 — COMMENTED** [review]
Looks good! A few comments

File: Library/Homebrew/dev-cmd/audit.rb
Comment (dtrodrigues) [reply Library/Homebrew/dev-cmd/audit.rb:921]: I had thought about unprivating and using that constant from `utils/pypi`, but I didn't want to include the entire file just for the constant.

File: Library/Homebrew/dev-cmd/audit.rb
Comment (dtrodrigues) [reply Library/Homebrew/dev-cmd/audit.rb:925]: We probably want to be more conservative and just do it for PyPI urls for now until a use case arises for other domains. `!=` does read more clearly than negating an XOR, so I'll make that change.

File: Library/Homebrew/dev-cmd/audit.rb
Comment (Rylan12) [reply Library/Homebrew/dev-cmd/audit.rb:921]: Is this a significant performance drop? I don't think the PyPI module is very large. In my opinion, it's better to have consistency and not repeat this hardcoded value anywhere.

Comment (dtrodrigues) [reply Library/Homebrew/dev-cmd/audit.rb]: You're right that it's not that big of a module, I'll go ahead and include and use it.

**Comment by fxcoudert** [comment]
I think this PR is misguided. The goal of the `checksum mismatch` audit is purely informative: it tells the maintainer that something _may_ be wrong, and that we have to check it and make a decision. As such, it is *expected* that it may be overridden from time to time, in justified cases. Therefore, suppressing the audit is not necessary.

I would also argue that suppressing the audit is a bad step: as a maintainer, I want to have this information. If a URL is changed, whether it's on PyPI or not, I'd rather be informed so I can check whether the content is actually the same. Imagine someone switches a PyPI URL to malicious content they host on github (or the other way around): I would rather be alerted of the issue, and check myself if there is a problem.

There are regularly packages that are updated without changing version number, and although I'd prefer it never happened, we have to live with that. Will we have to add an allow list, or adjust the logic of the code in brew every time? It does not seem desirable.

**Comment by dtrodrigues** [comment]
I agree that this is more of an informative audit vs a failure, but there isn't an option for a "non-failing warning" in the CI. PyPI URLs are a broad class where they could either be hosted on PyPI or GitHub, so an exception for it may make sense. Suppressing the fact that the domain changed isn't ideal, but neither is needing to merge a red PR.

If this could be done as a GitHub action label vs a CI check, that might be better, but I'm not sure how feasible that would be. It could even be more broad to cover domain changes in general, regardless of if the version changed.

@MikeMcQuaid indicated a preference for adjusting the audit logic here vs ignoring the audit failure, so I'll let him chime in with any additional perspective.

**Comment by fxcoudert** [comment]
If we want to never merge "red" CI, then this is one of the audits that need to be removed (or find a way to silence it manually). But that does not seem desirable at all to me…

**Comment by dtrodrigues** [comment]
Another possibility is to have CI pass a flag to the audit command when a manual "version checksum mismatch allowed" label is applied. It could then skip this check alone which would both give a green CI as well as flag it for reviewers to verify that the URL change makes sense.

**Comment by fxcoudert** [comment]
@dtrodrigues I am fine with that, but I think in the longer term there could be other “informative” audits of interest. We probably need to figure out a way to implement a two-level audit (errors and warnings)

File: Library/Homebrew/dev-cmd/audit.rb
Comment (Rylan12) [inline Library/Homebrew/dev-cmd/audit.rb:922]:
```suggestion
      if current_version == newest_committed_version &&
```

Another change that should be made here (unrelated). `current_version` and `previous_version` are always different so this will never be triggered anyway. Using `newest_committed_version` is correct because we want to make see if the version has changed since the last commit rather than the last version.

This will resolve #9083

Comment (Rylan12) [reply Library/Homebrew/dev-cmd/audit.rb:922]: Actually, that's incorrect. `current_version` and `previous_version` _can_ be the same but only if the `current_version` is also the first version (which is why the audit triggered for `vint` but not for `nim`

To avoid this, maybe we should set `previous_version`, `previous_version_scheme` and `previous_revision` to `nil` if there isn't a previous version.

File: Library/Homebrew/dev-cmd/audit.rb
Comment (MikeMcQuaid) [inline Library/Homebrew/dev-cmd/audit.rb:925]: I haven't really looked at the rest of the code but this doesn't seem to be anything specific to PyPI. If the URL changes: the checksum can change, too.

**Comment by MikeMcQuaid** [comment]
> If a URL is changed, whether it's on PyPI or not, I'd rather be informed so I can check whether the content is actually the same.

My take: if the URL changes, the checksum can change too. If the URL changes: the checksum doesn't need to change. If the checksum changes: the URL should (usually) change.

**Comment by MikeMcQuaid** [comment]
> Another possibility is to have CI pass a flag to the audit command when a manual "version checksum mismatch allowed" label is applied. It could then skip this check alone which would both give a green CI as well as flag it for reviewers to verify that the URL change makes sense.

Something like this makes a lot of sense as it's something that:
1) can only be done by maintainers
2) provides information logged on the PR about it happening and who did it

**Comment by MikeMcQuaid** [comment]
> @dtrodrigues I am fine with that, but I think in the longer term there could be other “informative” audits of interest. We probably need to figure out a way to implement a two-level audit (errors and warnings)

The only thing I can see that would be like a warning is have CI be green but have @BrewTestBot post a comment to the relevant pull request. Even then, though, I think I prefer the idea of an audit that's overridden with a specific label (rather than just "ignore audits").

**Comment by fxcoudert** [comment]
> audit that's overridden with a specific label

That sounds like a good solution

**Comment by dtrodrigues** [comment]
> I haven't really looked at the rest of the code but this doesn't seem to be anything specific to PyPI. If the URL changes: the checksum can change, too.

Would it make sense for the audit instead only to trigger if the URL is identical but the checksum changes?

**Comment by dtrodrigues** [comment]
Closed in favor of #9111

**Pull request closed without merging**
